# Module 4: Software for Data Analysis - Python Setup, Data Import, and Basic Exploration


Xiao Chen

Sept. 26, 2026

In this notebook, I will explore the Titanic dataset by examining its structure, data types, missing values, and basic numerical summaries. I will use these initial observations to identify questions that can be investigated through further analysis.

This notebook is organized in following steps:

1. Setup
2. Data Import
3. Initial Data Preview
4. Dataset Structure
5. Missing Values
6. Error-Reading Practice
7. Observations

# 1. Setup - Download the CSV from url.
Use the Titanic CSV dataset provided through the official pandas project:
https://raw.githubusercontent.com/pandas-dev/pandas/main/doc/data/titanic.csv

The dataset titanic.csv contains one row for each passenger and columns describing characteristics such as passenger class, age, fare, cabin, embarkation point, and survival status.

At this stage, we are only downloading the CSV file.

The code first checks whether the file is already available in the current Colab session. If it is, the download is skipped. If it is not, the file is downloaded from url.

### What Each Package Does

**`from pathlib import Path`**
* **What it is:** A built-in Python tool for handling folder locations and filenames.
* **Why use it:** Instead of worrying about slash directions (`/` vs `\`) across different operating systems (Windows vs. Mac/Linux), `Path` makes file paths safe, readable, and easy to join.


**`import pandas as pd`**
* **What it is:** The standard library for working with tabular data (spreadsheets).
* **Why use it:** Pandas allows us to load CSV files, clean missing values, filter rows, group numbers, and compute statistics. The `as pd` part is an industry-standard shortcut so we don't have to type `pandas` every time.


**`import urllib.request`**
* **What it is:** A Python module that provides tools for downloading and working with data from URLs.
* **Wh use it:** `urllib.request.urlretrieve()`  lets us download a file directly from a URL and save it to a specific location on your computer or notebook. In our example, we use it to download the Titanic CSV file from GitHub.

In [1]:
from pathlib import Path
import pandas as pd
import urllib.request

data_url = 'https://raw.githubusercontent.com/pandas-dev/pandas/main/doc/data/titanic.csv'

DATA_PATH = Path('/content/titanic.csv')

if not DATA_PATH.exists():
    urllib.request.urlretrieve(data_url, DATA_PATH)

print('CSV ready:', DATA_PATH.name)
print('File size (MB):', round(DATA_PATH.stat().st_size / 1_000_000, 1))

CSV ready: titanic.csv
File size (MB): 0.1


## 2. Data Import
- `read_csv()` reads a CSV into a table.
    - `keep_default_na=False` tells pandas not to automatically treat its usual missing-value markers (such as "NA" or "NaN") as missing.
    - `na_values=['']` tells pandas to treat empty strings as missing values (NaN).

Load the CSV file.

In [2]:
df = pd.read_csv(
    DATA_PATH,
    keep_default_na=False,
    na_values=['']
)
starting_rows = len(df)
print(f'Loaded {starting_rows:,} requests and {df.shape[1]} columns.')

Loaded 891 requests and 12 columns.


# 3. Initial Data Preview

- `head()` shows a few records; it cannot tell
us everything about a million-row dataset.

- `tail()` shows the last few records; it helps us see what is happening at the end of a large dataset.

- `concat()` combines multiple DataFrames together; it lets us view selected parts of a dataset in one table.

### 3A - Display the first 10 rows and the last 5 rows.


In [3]:
display(pd.concat([df.head(10), df.tail(5)]))

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S
5,6,0,3,"Moran, Mr. James",male,NaN,0,0,330877,8.4583,NaN,Q
6,7,0,1,"McCarthy, Mr. Timothy J",male,54.0,0,0,17463,51.8625,E46,S
7,8,0,3,"Palsson, Master Gosta Leonard",male,2.0,3,1,349909,21.0750,NaN,S
8,9,1,3,"Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)",female,27.0,0,2,347742,11.1333,NaN,S
9,10,1,2,"Nasser, Mrs. Nicholas (Adele Achem)",female,14.0,1,0,237736,30.0708,NaN,C


**Findings**:
- SibSp = number of siblings or spouses aboard the Titanic.
- Parch = number of parents or children aboard the Titanic.
- Each row appears to represent one passenger aboard the Titanic, including whether they survived and their passenger class.
- The columns store information such as the passenger's name, sex, age, family relationships(SibSp, Parch), ticket number, fare, cabin, and port of embarkation.

### 3B: Check dataset dimensions

Display the DataFrame’s shape. In a markdown cell, explain the number of rows, the number of columns, and what those two numbers mean in this dataset.


In [4]:
df.shape

(891, 12)

**Findings**:
- The DataFrame has 891 rows and 12 columns.
- The 891 rows represent individual passengers in the Titanic dataset, while the 12 columns represent different pieces of information recorded about each passenger, such as their survival status, name, age, sex, ticket, and fare.

### 3C: Review the column names

Display the complete list of column names. In a markdown cell, identify two columns containing categorical information, two containing numerical information, and one whose meaning was not immediately obvious to you.


In [5]:
df.columns.tolist()

['PassengerId',
 'Survived',
 'Pclass',
 'Name',
 'Sex',
 'Age',
 'SibSp',
 'Parch',
 'Ticket',
 'Fare',
 'Cabin',
 'Embarked']

**Findings**:
- Two columns containing categorical information are Sex and Embarked, because they contain categories such as male/female and different port codes.
- Two columns containing numerical information are Age and Fare, which contain passengers' ages and ticket prices.
- The Pclass column was not immediately obvious to me because it contains numbers (1, 2, and 3), but these numbers represent passenger classes rather than a continuous numerical measurement.

### 3D: Check data types

Display the data type assigned to every column. In a markdown cell, explain the difference between at least two data types shown in the output. Identify one integer column, one decimal or floating-point column, and one text column.

In [6]:
df.dtypes

,0
PassengerId,int64
Survived,int64
Pclass,int64
Name,object
Sex,object
Age,float64
SibSp,int64
Parch,int64
Ticket,object
Fare,float64


**Findings**:
- The DataFrame contains several different data types. An integer (int64) stores whole numbers, such as PassengerId, while a floating-point (float64) stores numbers that can contain decimal values, such as Fare and Age. A text (object) column stores text values, such as passenger names in the Name column.

# 4. Dataset Structure
- isna().sum() checks how many Passenger IDs are missing.
- duplicated().sum() checks how many Passenger IDs are repeated.

Display a structural summary.

Use info() to display a technical summary of the DataFrame. In a markdown cell, identify two useful pieces of information supplied by this output.


In [7]:
# info() reports column names, non-missing counts, and data types.
df.info()

print('Missing Passenger IDs:', df['PassengerId'].isna().sum())
print('Repeated Passenger IDs:', df['PassengerId'].duplicated().sum())


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB
Missing Passenger IDs: 0
Repeated Passenger IDs: 0


**Findings**:
- The info() output provides useful information about the number of non-null values in each column and the data type assigned to each column.
- For example, Age has 714 non-null values, meaning some ages are missing, and Fare has a float64 data type because it can contain decimal values.
- The output also shows that the DataFrame contains 891 rows and 12 columns.

# 5. Missing Value

- `isna()` asks whether a value is missing.
    - In this notebook, only blank cells are treated as missing; text values such as `NA` or `NaN` are not automatically treated as missing because of the `read_csv()` settings.

- `sum()` counts those True answers.
- The percentage is the count divided by the number of missings, multiplied by 100.
- `pd.DataFrame()` creates a summary table of missing-value counts and percentages.
- `sort_values()` orders the table from the most missing values to the least.





### 5A: Calculate and display the number of missing values in every column.

In a markdown cell, answer the following questions in 3–4 complete sentences:
- Which columns contain missing values?
- Which column contains the most missing values?
- Why should an analyst identify missing values before beginning a detailed analysis?

ps: calculate percentages.

In [8]:
missing_counts = df.isna().sum()
missing_percent = 100 * missing_counts / len(df)
missing_table = pd.DataFrame({
    'missing_count': missing_counts,
    'missing_percent': missing_percent.round(2)
})
display(missing_table.sort_values('missing_count', ascending=False))

,missing_count,missing_percent
Cabin,687,77.10
Age,177,19.87
Embarked,2,0.22
PassengerId,0,0.00
Name,0,0.00
Pclass,0,0.00
Survived,0,0.00
Sex,0,0.00
Parch,0,0.00
SibSp,0,0.00


**Findings**:
- The columns containing missing values are Cabin, Age, and Embarked.
- The Cabin column contains the most missing values, with 687 missing entries, or 77.10% of the dataset.
- An analyst should identify missing values before beginning a detailed analysis because they can affect calculations and lead to incomplete or misleading results. Understanding where missing values occur also helps an analyst decide how to handle them appropriately.

### 5B. Produce a basic numerical summary

Use describe() to summarize the numerical columns. In a markdown cell, identify two statistics in the output and explain what each one communicates.


In [9]:
df.describe()

,PassengerId,Survived,Pclass,Age,SibSp,Parch,Fare
count,891.000000,891.000000,891.000000,714.000000,891.000000,891.000000,891.000000
mean,446.000000,0.383838,2.308642,29.699118,0.523008,0.381594,32.204208
std,257.353842,0.486592,0.836071,14.526497,1.102743,0.806057,49.693429
min,1.000000,0.000000,1.000000,0.420000,0.000000,0.000000,0.000000
25%,223.500000,0.000000,2.000000,20.125000,0.000000,0.000000,7.910400
50%,446.000000,0.000000,3.000000,28.000000,0.000000,0.000000,14.454200
75%,668.500000,1.000000,3.000000,38.000000,1.000000,0.000000,31.000000
max,891.000000,1.000000,3.000000,80.000000,8.000000,6.000000,512.329200


**Findings**:
- The mean represents the average value of a numerical column; for example, the average age of passengers with recorded ages was approximately 29.7 years.
- The 50% statistic represents the median, meaning half of the recorded values are below this value and half are above it; for Age, the median was 28 years.
- These statistics help provide a quick understanding of the typical values in the dataset.

# 6. Error-Reading Practice

Create a code cell containing an intentional mistake by using a DataFrame method name that does not exist, such as:
- titanic.headd()
- Run the cell and examine the error message. Do not delete the error output.
- In the next markdown cell, answer each question in 1–2 complete sentences:
  - What type of error was reported?
  - Which line of code caused the error?
  - What part of the message helped you identify the problem?
  - How did you correct the code?

Create a new code cell containing the corrected command and run it successfully.



In [10]:
titanic.headd()

NameError: name 'titanic' is not defined

**QAs**:

What type of error was reported?
- A NameError was reported because Python could not find a variable named titanic.

Which line of code caused the error?
- The line titanic.headd() caused the error.

What part of the message helped you identify the problem?
- The message name 'titanic' is not defined showed that titanic was not a defined variable in the notebook.

How did you correct the code?
- I used the correct DataFrame name, df, and corrected the method name from headd() to head().

In [11]:
df.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


# 7.  Initial Data Observations

Add a markdown section titled “Initial Data Observations.” Write five numbered observations in complete sentences. Include:
- One observation about the dataset’s size
- One observation about the columns or data types
- One observation about missing values
- One observation based on the numerical summary
- One question that could be investigated through additional analysis

Your observations must describe only what the notebook output supports. Do not make causal claims or assume that a pattern exists without calculating or displaying evidence.

**Observations**:

- The dataset contains 891 rows and 12 columns, with each row representing a passenger record.

- The DataFrame contains five integer columns, two floating-point columns, and five object columns.

- Missing values are present in Age, Cabin, and Embarked, with Cabin having the most missing values at 687 entries (77.10%).

- The numerical summary shows that the median passenger age is 28 years, while the mean age among passengers with recorded ages is approximately 29.7 years.

- One question for additional analysis is: How does the number of passengers who survived compare across the different passenger classes (Pclass)?